### PR TITLE
[AIE2] Enable LDA to use optimal itinerary

### PR DESCRIPTION
**Pull request by krishnamtibrewala**
Enable LDA , and PostInc. Load with various addr mode to use optimal itinerary

### CODE REVIEW

File: llvm/lib/Target/AIE/AIE2Schedule.td [hunk 1]
```diff
@@ -462,11 +468,35 @@ MemInstrItinData<II_LDA_R,
 	EmptyCycles<4>, SimpleCycle<R_WA_PORT>], [7,1,1],
     MemoryCycles<[5]>>,
 MemInstrItinData<II_LDA_POST_1D,
-    [AvoidPartWordStore, EmptyCycles<1>, SimpleCycle<LOAD_UNIT_A>, EmptyCycles<4>,
+    [AvoidPartWordStore, SimpleCycle<P_WM_PORT>, SimpleCycle<LOAD_UNIT_A>, EmptyCycles<4>,
```
Comment (krishnamtibrewala) [inline llvm/lib/Target/AIE/AIE2Schedule.td:471]: Hi @gbossu, @andcarminati , please review the modified schedule for II_LDA_POST_1D,
based on my understanding of ISA we will update the pointer reg in E1 stage and thus will use the write port in E1 also.

File: llvm/lib/Target/AIE/AIE2Schedule.td [hunk 2]
```diff
@@ -498,11 +504,36 @@ MemInstrItinData<II_LDA_POST_1D_R,
 	EmptyCycles<4>, SimpleCycle<R_WA_PORT>], [7,1,1,1],
     MemoryCycles<[5]>>,
 MemInstrItinData<II_LDA_POST_2D,
-    [AvoidPartWordStore, EmptyCycles<1>, SimpleCycle<LOAD_UNIT_A>, EmptyCycles<4>,
+    [AvoidPartWordStore, PrefixCycle<DC_WM_PORT>, SimpleCycle<P_WM_PORT>, SimpleCycle<LOAD_UNIT_A>,
+  	 EmptyCycles<4>,
```
Comment (krishnamtibrewala) [inline llvm/lib/Target/AIE/AIE2Schedule.td:514]: Please verify this too.

File: llvm/lib/Target/AIE/AIE2Schedule.td [hunk 3]
```diff
@@ -535,11 +541,35 @@ MemInstrItinData<II_LDA_POST_2D_R,
 	EmptyCycles<4>, SimpleCycle<R_WA_PORT>], [7,1,1,1,1],
     MemoryCycles<[5]>>,
 MemInstrItinData<II_LDA_POST_3D,
-    [AvoidPartWordStore, EmptyCycles<1>, SimpleCycle<LOAD_UNIT_A>, EmptyCycles<4>,
+    [AvoidPartWordStore, PrefixCycle<DC_WM_PORT>, SimpleCycle<P_WM_PORT>, SimpleCycle<LOAD_UNIT_A>, EmptyCycles<4>,
```
Comment (krishnamtibrewala) [inline llvm/lib/Target/AIE/AIE2Schedule.td:544]: Please verify this too.

File: llvm/include/llvm/Target/Target.td [hunk 4]
```diff
@@ -583,11 +583,15 @@ class EncodingByHwMode<list<HwMode> Ms = [], list<InstructionEncoding> Ts = []>
   list<InstructionEncoding> Objects = Ts;
 }
 
+class OperandRegClass<int OperIdx, RegisterClass Reg> {
```
Comment (gbossu) [inline llvm/include/llvm/Target/Target.td]: Note: I think this should rather be in `llvm/Target/AIETarget.td`

Comment (krishnamtibrewala) [reply llvm/include/llvm/Target/Target.td]: We needed  `list<ItinRegClassPair>` as part of `class Instruction` thus we needed this to be added to Target.td

File: llvm/lib/Target/AIE/AIE2GenInstrInfo.td [hunk 5]
```diff
@@ -201,12 +202,16 @@ let mayLoad = true, mayStore = true in {
...
   let Itinerary = II_LDA in {
-  def LDA_dms_lda_idx : AIE2_dms_lda_inst_lda< (outs OP_mLdaScl:$mLdaScl) , (ins eP:$ptr, eDJ:$dj),
-      "lda", "$mLdaScl, [$ptr, $dj]">, 
-    AIE2_ag_all__ag_idx;
-  def LDA_dms_lda_idx_imm : AIE2_dms_lda_inst_lda< (outs OP_mLdaScl:$mLdaScl) , (ins eP:$ptr, imm6x4:$imm),
-      "lda", "$mLdaScl, [$ptr, $imm]">, 
-    AIE2_ag_all__ag_idx_imm;
+  let ItineraryRegPairs = [ItinRegClassPair<II_LDA_R,[OperandRegClass<0, eR>]>, ItinRegClassPair<II_LDA_P,[OperandRegClass<0, eP>]>, 
+                           ItinRegClassPair<II_LDA_M,[OperandRegClass<0, eM>]>, ItinRegClassPair<II_LDA_DC,[OperandRegClass<0, eDC>]>,
+                           ItinRegClassPair<II_LDA_DJ,[OperandRegClass<0, eDJ>]>, ItinRegClassPair<II_LDA_DN,[OperandRegClass<0, eDN>]>] in {
```
Comment (gbossu) [inline llvm/lib/Target/AIE/AIE2GenInstrInfo.td]: I think LDA can also write to `lr`. I guess in that case we'll be conservative and book all ports?

Comment (krishnamtibrewala) [reply llvm/lib/Target/AIE/AIE2GenInstrInfo.td]: Yes, anytime any of `ItineraryRegPairs` is not matched we fall back to the conservative and reserved according to default Itinerary, in this case `II_LDA`

File: llvm/lib/Target/AIE/AIE2Schedule.td [hunk 6]
```diff
@@ -495,11 +501,36 @@ MemInstrItinData<II_LDA_POST_1D_R,
...
+    [AvoidPartWordStore, EmptyCycles<1>, SimpleCycle<LOAD_UNIT_A>,
+	EmptyCycles<4>, SimpleCycle<DN_WM_PORT>], [7,1,1,1,1],
+    MemoryCycles<[5]>>,
+MemInstrItinData<II_LDA_POST_2D_M,
+    [AvoidPartWordStore, EmptyCycles<1>, SimpleCycle<LOAD_UNIT_A>,
+	EmptyCycles<4>, SimpleCycle<M_WM_PORT>], [7,1,1,1,1],
+    MemoryCycles<[5]>>,
+MemInstrItinData<II_LDA_POST_2D_P,
+    [AvoidPartWordStore, EmptyCycles<1>, SimpleCycle<LOAD_UNIT_A>,
+	EmptyCycles<4>, PrefixCycle<DC_WM_PORT>, SimpleCycle<P_WM_PORT>], [7,1,1,1,1],
```
Comment (gbossu) [inline llvm/lib/Target/AIE/AIE2Schedule.td]: Why do we book DC and P ports here?

Comment (gbossu) [reply llvm/lib/Target/AIE/AIE2Schedule.td]: (Same for 3D post-inc)

Comment (krishnamtibrewala) [reply llvm/lib/Target/AIE/AIE2Schedule.td]: Its a mistake, there is no reason to reserve DC port. Will fix it

File: llvm/include/llvm/Target/Target.td (hunk 4 above)
Comment (gbossu) [reply llvm/include/llvm/Target/Target.td]: We probably could have included `Target.td` in `AIETarget.td`

File: llvm/lib/Target/AIE/AIE2GenInstrInfo.td [hunk 7]
```diff
@@ -228,16 +228,20 @@ let mayLoad = true, mayStore = false in {
   }
 
   let Itinerary = II_LDA_POST_2D, Constraints = "$ptr = $ptr_out" in {
-    let DecoderMethod = "DecodeLDA_2D_Instruction" in {
+    let ItineraryRegPairs = [ItinRegClassPair<II_LDA_POST_2D_R,[OperandRegClass<0, eR>]>, ItinRegClassPair<II_LDA_POST_2D_P,[OperandRegClass<0, eP>]>, 
+                             ItinRegClassPair<II_LDA_POST_2D_M,[OperandRegClass<0, eM>]>, ItinRegClassPair<II_LDA_POST_2D_DC,[OperandRegClass<0, eDC>]>,
+                             ItinRegClassPair<II_LDA_POST_2D_DJ,[OperandRegClass<0, eDJ>]>, ItinRegClassPair<II_LDA_POST_2D_DN,[OperandRegClass<0, eDN>]>],
```
Comment (gbossu) [inline llvm/lib/Target/AIE/AIE2GenInstrInfo.td]: Super-nit: I think one itinerary per line would be much more readable.
(No need to address this)

**Review by gbossu — DISMISSED** [review]
LGTM

**Review by gbossu — APPROVED** [review]

**Pull request merged**